mirrord/layer/src/socket.rs

```diff
@@ -379,25 +385,110 @@ impl OutgoingSelector {
                     || (select_address.port() == 0 && select_address.ip() == address.ip())
                     || select_address == address
             }
-            // TODO(alex): DNS may not trigger the filter, as the local resolved address may be
-            // different from the remote resolved.
-            AddressFilter::Name((ref name, port)) => format!("{name}:{port}")
-                .to_socket_addrs()
-                .map(|mut resolved| {
-                    resolved.any(|resolved_address| {
-                        resolved_address == SocketAddr::new(address.ip(), 0)
-                    })
-                })
-                .unwrap_or_default(),
+            // TODO(alex): We could enforce this at the type level, by converting `OutgoingWhatever`
+            // to a type that doesn't have `AddressFilter::Name`.
+            AddressFilter::Name(_) => unreachable!("BUG: We skip these in the iterator!"),
             AddressFilter::Subnet((subnet, port)) => {
                 subnet.contains(&address.ip()) && (port == 0 || port == address.port())
             }
         };
 
-        match self {
+        let resolved_hosts = match &self {
+            OutgoingSelector::Unfiltered => HashSet::default(),
+            OutgoingSelector::Remote(_) => self.resolve_dns::<true>()?,
+            OutgoingSelector::Local(_) => self.resolve_dns::<false>()?,
+        };
+        let hosts = resolved_hosts.iter();
+
+        Ok(match self {
             OutgoingSelector::Unfiltered => true,
-            OutgoingSelector::Remote(list) => list.iter().filter(filter_protocol).any(any_address),
-            OutgoingSelector::Local(list) => !list.iter().filter(filter_protocol).any(any_address),
+            OutgoingSelector::Remote(list) => list
+                .iter()
+                .chain(hosts)
+                .filter(skip_unresolved)
+                .filter(filter_protocol)
+                .any(any_address),
+            OutgoingSelector::Local(list) => !list
+                .iter()
+                .chain(hosts)
+                .filter(skip_unresolved)
```

Comment:
```suggestion
                .filter(skip_unresolved)
                .chain(hosts)
```
Also here if possible.